sql/analyzer/hoist_select_exists.go

```diff
@@ -0,0 +1,162 @@
+			right = p.Child
+		}
+
+		if gb, ok := right.(*plan.GroupBy); ok {
+			right = gb.Child
+		}
+
+		outerFilters, _, err := FixFieldIndexesOnExpressions(ctx, scope, a, append(left.Schema(), right.Schema()...), outerFilters...)
+		if err != nil {
+			return n, transform.SameTree, err
+		}
+
+		switch joinType {
+		case plan.AntiJoinType:
+			return plan.NewAntiJoin(left, right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
+		case plan.SemiJoinType:
+			return plan.NewSemiJoin(left, right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
+		default:
+			panic("expected SemiJoinType or AntiJoinType")
+		}
+	})
+}
+
+// pluckCorrelatedExistsSubquery scans a filter for [note] WHERE EXISTS, and then attempts to
+// extract the subquery, correlated filters, a modified outer scope (net subquery and filters),
+// and the new target joinType
+func pluckCorrelatedExistsSubquery(filter *plan.Filter) (sql.Node, sql.Node, []sql.Expression, plan.JoinType) {
+	// if filter has a correlated exists, we remove it from the filter and return the new sq and join condition
+	var decorrelated sql.Node
+	var outerFilters []sql.Expression
+
+	filters := splitConjunction(filter.Expression)
+	var newFilters []sql.Expression
+	var joinType plan.JoinType
+	for _, f := range filters {
+		switch e := f.(type) {
+		case *plan.ExistsSubquery:
+			decorrelated, outerFilters = decorrelateOuterCols(e.Query, len(filter.Schema()))
+			if len(outerFilters) == 0 {
+				return filter, nil, nil, plan.UnknownJoinType
+			}
+			joinType = plan.SemiJoinType
+		case *expression.Not:
+			esq, ok := e.Child.(*plan.ExistsSubquery)
+			if !ok {
+				return filter, nil, nil, plan.UnknownJoinType
+			}
+			decorrelated, outerFilters = decorrelateOuterCols(esq.Query, len(filter.Schema()))
+			if len(outerFilters) == 0 {
+				return filter, nil, nil, plan.UnknownJoinType
+			}
+			joinType = plan.AntiJoinType
+		default:
+		}
+	}
+	if len(outerFilters) == 0 {
+		return nil, nil, nil, plan.UnknownJoinType
+	}
+	if len(newFilters) == 0 {
+		return filter.Child, decorrelated, outerFilters, joinType
+	}
+	newFilter := plan.NewFilter(expression.JoinAnd(newFilters...), filter.Child)
+	return newFilter, decorrelated, outerFilters, joinType
+}
+
+// decorrelateOuterCols returns an optionally modified subquery and extracted
+// filters referencing an outer scope.
+func decorrelateOuterCols(e *plan.Subquery, scopeLen int) (sql.Node, []sql.Expression) {
+	var outerFilters []sql.Expression
+	n, same, _ := transform.Node(e.Query, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		f, ok := n.(*plan.Filter)
+		if !ok {
+			return n, transform.SameTree, nil
+		}
+		filters := splitConjunction(f.Expression)
+		var newFilters []sql.Expression
+		for _, f := range filters {
+			var outerRef bool
+			transform.InspectExpr(f, func(e sql.Expression) bool {
+				gf, ok := e.(*expression.GetField)
+				if !ok {
+					return false
+				}
+				if gf.Index() < scopeLen {
```

Comment:
This isn't a scope len, as used

And in fact you aren't passing the scope from the top-level rule down into these functions that would need it

Why does this work? Are you missing a test case?

Reply:
good catch, subquery expression, added test